Change the shipping rate API response and request args

nima-karimi opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

Implements part of #1117. Based on #1171 and #1192

This PR changes both the response and request arguments for the shipping rate API along with the batch API. Some of the prominent changes include:

- Accept the country code as part of the shipping rate data structure in requests instead of as a separate argument
- Rename the `country_code` parameter in both the response and request to `country` for **ALL shipping rates endpoints**. This is to align it with how it's represented internally and in the DB.
- Return the shipping rate ID in the response. This is useful because we are gradually moving toward having multiple rates per country and relying on the country_code as an identifier isn't gonna work anymore.
- Change the `DELETE` endpoint to only work with IDs instead of country codes (because of the above).
- Add an edit rate endpoint to allow editing a shipping rate by its ID

The request schema for creating a shipping rate is modified to the following (note that the `country_code` parameter is renamed to `country` and it's moved inside the rate data structure):

`POST mc/shipping/rates`
```JSON
{
      "country": "PT",
      "method": "flat_rate",
      "currency": "USD",
      "rate": 80,
      "options": {
        "free_shipping_threshold": "200"
      }
}
```

The batch create API request schema is also changed to the following:

`POST mc/shipping/rates/batch`
```JSON
{
  "rates": [
    {
      "country": "PT",
      "method": "flat_rate",
      "currency": "USD",
      "rate": 80,
      "options": {
        "free_shipping_threshold": "2000"
      }
    },
    {
      "country": "DE",
      "method": "flat_rate",
      "currency": "USD",
      "rate": 50,
      "options": {
        "free_shipping_threshold": "1200"
      }
    },
    {
      "country": "US",
      "method": "flat_rate",
      "currency": "USD",
      "rate": 10
    }
  ]
}
```

---

The response for getting all the available shipping rates is an array of shipping rate objects (without any `rates` parameter):

`GET mc/shipping/rates/`
```JSON
[
    {
        "id": "21",
        "country": "DE",
        "method": "flat_rate",
        "currency": "USD",
        "rate": "50",
        "options": {
            "free_shipping_threshold": 1200
        }
    },
    {
        "id": "33",
        "country": "PT",
        "method": "flat_rate",
        "currency": "USD",
        "rate": "80",
        "options": {
            "free_shipping_threshold": 2000
        }
    },
    {
        "id": "32",
        "country": "US",
        "method": "flat_rate",
        "currency": "USD",
        "rate": "10"
    }
]
```

---

The shipping rates suggestion API response is also modified to reflect the changes in the other endpoints. It accepts an array of country codes using the `country_codes` parameter and returns suggestions for all of them.

`GET mc/shipping/rates/suggestions?country_codes[]=ES&country_codes[]=US`

```JSON
[
    {
        "country": "ES"
        "method": "free_shipping",
        "currency": "USD",
        "rate": 0,
        "options": {
            "free_shipping_threshold": 100
        }
    },
    {
        "country": "US",
        "method": "flat_rate",
        "currency": "USD",
        "rate": 20,
        "options": {
            "free_shipping_threshold": 40.99
        }
    }
]
```

---

A new edit API is added that can update the rates using their ID:

`PUT/PATCH/POST mc/shipping/rates/33`

Request:
```JSON
{
      "country": "PT",
      "method": "flat_rate",
      "currency": "USD",
      "rate": 250,
      "options": {
        "free_shipping_threshold": "700"
      }
}
```

The response is an empty `HTTP 204 No Content`.

---

The delete endpoint now only works with an ID:

`DELETE mc/shipping/rates/33`

```JSON
{
    "status": "success",
    "message": "Successfully deleted rate."
}
```

The batch delete endpoint is similar to the normal endpoint but it can accept an array of `ids` as the parameter.

---

### Detailed test instructions:
<!-- Add detailed instructions for how to test that this PR fixes the issue and confirm that it doesn't break any other features :) -->

1. Test the APIs mentioned above using the correct request data structure and confirm that the response is as expected.


### Additional details:

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with `(Fix|Add|Tweak|Update) - `, for example:
> Fix - I took care of something that wasn't working.
> Add - I added something new that's pretty cool.
> Tweak - I made a small change.
> Update - I made big changes to something that wasn't broken.

Leave the "Changelog entry" header in place completely empty, without any summary if no changelog entry is needed.
If you remove the "Changelog entry" header, the title of Pull Request will be used as the changelog entry.
-->
### Changelog entry

> Update - Modify the shipping rates API request arguments and response schema

Commit:
Change the shipping rate API response and request args

Commit:
Fix the argument parameter key

Commit:
Update error message

Commit:
Remove shipping class rates from suggestions

mikkamp commented on src/API/Site/Controllers/MerchantCenter/ShippingRateBatchController.php:
```diff
@@ -17,9 +17,6 @@
  */
 class ShippingRateBatchController extends ShippingRateController {
 
-	use BatchSchemaTrait;
```
The import line for the BatchSchemaTrait at the top of the file can be removed as well.

mikkamp commented on src/API/Site/Controllers/MerchantCenter/ShippingRateController.php:
```diff
@@ -166,7 +158,24 @@ protected function get_suggestions_callback(): callable {
+					// Suggestions don't have a id.
+					$data = $response->get_data();
+					unset( $data['id'] );
+					$response->set_data( $data );
```
I haven't seen other examples of this in our controllers. When we register the route we provide the same schema as for a regular rate, so that suggests that the response would contain an ID. But then here we override the response and remove it.

Is there a cleaner way to handle this? Do we need to provide a custom schema when registering the route? What about using the filter in `prepare_item_for_response` would that be a cleaner way to remove the ID? Or do we need to make the base controller functions more flexible to allow different types of schema per request?

nima-karimi replied:
I should've done this in the beginning but given that the suggestion API seems to be getting very different from the CRUD API (both in the request arguments and the response schema), I've decided to give it its own home at `ShippingRateSuggestionsController`. Now it has its own schema and everything.

mikkamp commented on src/API/Site/Controllers/MerchantCenter/ShippingRateController.php:
```diff
@@ -304,69 +328,21 @@ protected function get_delete_rate_callback(): callable {
-
-		return array_map( [ $this, 'prepare_country_rates_for_output' ], $rates );
+		return $this->create_query()
+					->set_limit( 200 )
```
What is this limit connected to? Do we only want to handle a certain amount of shipping rates? Just wondering why we doubled the amount here.

nima-karimi replied:
No reason really. I'm not sure why it was there and there was no comment so I tried not to touch it the first time. But then realized we can have at least as many shipping rates as the number of countries to I increased to more than 100. But you're right there's no logic to it so I've now just removed it so that it won't confuse the next person who looks at it in the future 😅

mikkamp commented on src/API/Site/Controllers/MerchantCenter/ShippingRateController.php:
```diff
@@ -386,80 +362,67 @@ protected function create_query(): ShippingRateQuery {
+				'context'              => [ 'view', 'edit' ],
+				'validate_callback'    => 'rest_validate_request_arg',
+				'default'              => [],
+				'properties'           => [
```
Won't we have additional options here like the shipping classes in some cases? Should we include that?

nima-karimi replied:
No that's now removed because we no longer want to allow the user to edit their shipping class rates manually. It's done all using the Automatic shipping rate syncing method introduced in https://github.com/woocommerce/google-listings-and-ads/pull/1192

mikkamp commented on src/API/Site/Controllers/MerchantCenter/ShippingRateController.php:
```diff
@@ -175,25 +184,51 @@ protected function get_suggestions_callback(): callable {
+	 */
+	protected function get_update_rate_callback(): callable {
+		return function ( Request $request ) {
+			$id = (string) $request->get_param( 'id' );
```
Is there a particular reason we need to cast it to a string (same is done for the delete callback)? In the schema it's a number and in the DB it's an int, wouldn't it be better to modify the function `get_shipping_rate_by_id` to accept an int?

nima-karimi replied:
It's a workaround to prevent any issue with PHP not recognizing `bigint`. That's a very unlikely scenario because we usually wouldn't save a lot of rates in the DB but just to be on the safe side I decided to deal with all IDs as strings. I think the schema uses `is_numeric` to check the value so even a string representation of a number would pass.

mikkamp replied:
Interesting, I don't recall taking this into consideration for other tables. I always had the impression since WP uses a bigint for storing post ID's and treats it in PHP as an int we should be alright to do the same for our other tables. WP tables use an even larger range since it's an unsigned bigint. I guess in theory it could be a problem for some platforms that run on 32 bit (but probably not very common).

mikkamp reviewed (commented):
The changes look good and it looks clearer this way. Running the tests I get all the results as expected. I just left a few comments about some of the code, but none of those are blocking.

Commit:
Create separate controller for shipping rate suggestions

Commit:
Remove the query limit

nima-karimi commented:
Thanks, Mik for your review. Ready for another round 🙏

nima-karimi commented:
Based on the discussion with @ecgan on p1642498006075400-slack-C01E9LB4X61, we decided to change the output of the suggestions API (`GET mc/shipping/rates/suggestions`) and merge all the suggested rate arrays into one. So now I've removed the batch suggestions API and modified the response of the suggestions API to return an array of suggested shipping rates.

The input parameter for the `GET mc/shipping/rates/suggestions` API is now an array of `country_codes`.

Commit:
Return empty list of suggestions instead of an error

mikkamp commented on src/API/Site/Controllers/MerchantCenter/ShippingRateSuggestionsController.php:
```diff
@@ -0,0 +1,148 @@
+use Automattic\WooCommerce\GoogleListingsAndAds\Proxies\RESTServer;
+use Automattic\WooCommerce\GoogleListingsAndAds\Shipping\ShippingZone;
+use WP_REST_Request as Request;
+use WP_REST_Response as Response;
```
This class is no longer used now that we don't return a custom response for an empty list of rates.

mikkamp commented on src/API/Site/Controllers/MerchantCenter/ShippingRateSuggestionsController.php:
```diff
@@ -0,0 +1,148 @@
+ *
+ * @since x.x.x
+ */
+class ShippingRateSuggestionsController extends BaseController implements ISO3166AwareInterface {
```
The ISO3166AwareInterface was used for converting a country code to country name, we can remove that from this class and from the ShippingRateController as well. Also the import line at the top of the file can be removed.

mikkamp replied:
Actually it seems like we also use it for validating the country code, so you can ignore this comment.

mikkamp reviewed (approved):
Thanks for the changes, it's looking good and still runs well.
Just left a few small comments, but none of those are blocking so I'll go ahead and approve the PR.

Commit:
Remove unused import

Commit:
Merge branch 'feature/add-automatic-setting-for-shipping' into feature/modify-shipping-rate-api

# Conflicts:
#	src/API/Site/Controllers/MerchantCenter/ShippingRateController.php
